src/encoded/ingestion/structured_data.py

```diff
@@ -0,0 +1,606 @@
+
+    def _load_json_file(self, file: str) -> None:
+        with open(file) as f:
+            self._add(get_type_name(file), json.load(f))
+
+    def _load_reader(self, reader: RowReader, type_name: str) -> None:
+        schema = None
+        noschema = False
+        structured_column_data = _StructuredRowData(reader.header)
+        for row in reader:
+            if not schema and not noschema and not (schema := Schema.load_by_name(type_name, portal=self._portal)):
+                noschema = True  # Create schema here just so we don't create it if there are no rows.
+            structured_row = structured_column_data.create_row()
+            for column_name, value in row.items():
+                structured_column_data.set_value(structured_row, column_name, value, schema, reader.location)
+            self._add(type_name, structured_row)
+
+    def _add(self, type_name: str, data: Union[dict, List[dict]]) -> None:
+        if self._prune:
+            remove_empty_properties(data)
+        if type_name in self.data:
+            self.data[type_name].extend([data] if isinstance(data, dict) else data)
+        else:
+            self.data[type_name] = [data] if isinstance(data, dict) else data
+
+    def _note_issues(self, issues: Optional[List[str]], source: str) -> None:
+        if issues:
+            if not self._issues:
+                self._issues = []
+            self._issues.append({source: issues})
+
+
+class _StructuredRowData:
+
+    def __init__(self, column_names: List[str]) -> None:
+        self._row_template = self._parse_into_row_template(column_names)
+
+    def create_row(self) -> dict:
+        return copy.deepcopy(self._row_template)
+
+    @staticmethod
+    def set_value(row: dict, column_name: str, value: str, schema: Optional[Schema], loc: int) -> None:
+
+        def setv(row: Union[dict, list], column_name_components: List[str], parent_array_index: int = -1) -> None:
+
+            if not column_name_components:
+                return
+            if isinstance(row, list):
+                if parent_array_index < 0:
+                    for row_item in row:
+                        setv(row_item, column_name_components)
+                else:
+                    setv(row[parent_array_index], column_name_components)
+                return
+            if not isinstance(row, dict):
+                return
+
+            column_name_component = column_name_components[0]
+            array_name, array_index = _StructuredRowData._get_array_info(column_name_component)
+            name = array_name if array_name else column_name_component
+            if len(column_name_components) > 1:
+                if not isinstance(row[name], dict) and not isinstance(row[name], list):
+                    row[name] = {}
+                setv(row[name], column_name_components[1:], parent_array_index=array_index)
+                return
+
+            nonlocal column_name, value, schema, loc
```

Comment:
Neat usage, though I think you only need `nonlocal column_name` since this is the only ref outside the nested function?

Reply:
No I don't think so; value, schema, and loc are used as well (unless I'm misunderstanding your question).